state-sync/data-streaming-service/src/data_stream.rs

```diff
@@ -263,14 +263,25 @@ impl<T: AptosDataClientInterface + Send + Clone + 'static> DataStream<T> {
         &mut self,
         global_data_summary: &GlobalDataSummary,
     ) -> Result<(), Error> {
-        // Determine how many requests (at most) can be sent to the network
-        let num_sent_requests = self.get_sent_data_requests()?.len() as u64;
-        let max_concurrent_requests = self.get_max_concurrent_requests();
-        let max_num_requests_to_send = max_concurrent_requests
-            .checked_sub(num_sent_requests)
-            .ok_or_else(|| {
-                Error::IntegerOverflow("Max number of requests to send has overflown!".into())
-            })?;
+        // Calculate the number of in-flight requests (i.e., requests that haven't completed)
+        let num_pending_requests = self.get_num_pending_data_requests()?;
+        let num_complete_pending_requests = self.get_num_complete_pending_requests()?;
+        let num_in_flight_requests =
+            num_pending_requests.saturating_sub(num_complete_pending_requests);
+
+        // Calculate the max number of requests that can be sent now
+        let max_pending_requests = self.streaming_service_config.max_pending_requests;
+        let max_num_requests_to_send = if num_pending_requests >= max_pending_requests {
+            0 // We're already at the max number of pending requests (don't do anything)
```

Comment:
the `else` would still return 0 in this case, do we need this separate case?

Reply:
Yeah, good question. 😄 I think this is probably a personal thing (i.e., being explicit with the check instead of relying on the saturating sub to handle the underflow). Personally, I find it more readable, and it'll mean that when I come to change this code in the future, I don't forget about the edge case 😆 I'm gonna leave it as-is for now, but if you feel strongly, let me know and I can merge them.